File: flag.lua

-- mimic https://golang.org/pkg/flag/
local flag = {}

flag.defined = {}
flag.args = {}


local function info(...)
    print(...)
end

local function warn(...)
    print(...)
end

local function fatal(...)
    print(...)
    os.exit(1)
end

local function convert(t, val)
    if t == "string" then
        return val
    elseif t == "bool" then
        return true
    elseif t == "number" then
        return tonumber(val)
    else
        fatal("unknown type", t)
    end
end

--
-- syntax:
-- -flag=x
--
-- -flag
-- -flag x
function flag.Parse(...)
    local args = {...}
    local argc = select("#", ...)

    for i = 1, argc do
        local arg = args[i]
        if arg:sub(1, 1) ~= "-" then
            table.insert(flag.args, arg)
        else
            local k, v = arg:match("%-(.+)=(.+)")
            if k then
                -- -flag=x
                local entry = flag.defined[k]
                if not entry then
                    fatal("undefined flag", k)
                else
                    if entry.hasDefault then
                        table.remove(entry.value)
                        entry.hasDefault=false
                    end
                    table.insert(entry.value, convert(entry.kind, v))
                end
            else
                k = arg:match("%-(.+)")
                if k then
                    local a = args[i+1]
                    if a and a:sub(1, 1) ~= "-" then
                        -- -flag x
                        local entry = flag.defined[k]
                        if not entry then
                            fatal("undefined flag", k)
                        else
                            if entry.hasDefault then
                                table.remove(entry.value)
                                entry.hasDefault=false
                            end
                            table.insert(entry.value, convert(entry.kind, a))
                        end
                        i = i + 1
                    else
                        -- -flag
                        local entry = flag.defined[k]
                        if not entry then
                            fatal("undefined flag", k)
                        else
                            if entry.kind ~= "bool" then
                                fatal("-" .. k, "need one argument")
                            end
                            if entry.hasDefault then
                                table.remove(entry.value)
                                entry.hasDefault=false
                            end
                            table.insert(entry.value, convert(entry.kind))
                        end
                    end
                end
            end
        end
    end
end


function flag.Args()
    return flag.args
end

function flag.Bool(name, value, usage)
    if type(value) ~= "boolean" then
        fatal("invalid value in flag.Bool")
    end
    local ptr = {value}
    flag.defined[name] = {
        hasDefault = true,
        ["default"] = value,
        ["usage"] = usage, 
        ["kind"] = "bool",
        ["value"] = ptr,
    }
    return ptr
end

function flag.String(name, value, usage)
    if type(value) ~= "string" then
        fatal("invalid value in flag.String")
    end
    local ptr = {value}
    flag.defined[name] = {
        hasDefault = true,
        ["default"] = value,
        ["usage"] = usage, 
        ["kind"] = "string", 
        ["value"] = ptr,
    }
    return ptr
end

function flag.Number(name, value, usage)
    if type(value) ~= "number" then
        fatal("invalid value in flag.Number")
    end
    local ptr = {value}
    flag.defined[name] = {
        hasDefault = true,
        ["default"] = value,
        ["usage"] = usage, 
        ["kind"] = "number",
        ["value"] = ptr,
    }
    return ptr
end

function flag.PrintDefaults()
    -- -x int
    --    usage-message-for-x (default 7)
    for name, entry in pairs(flag.defined) do
        info(string.format("-%s %s", name, entry.kind))
        info(string.format("  %s (default %s)", entry.usage, tostring(entry.default)))
    end
end

return flag

File: flag.ts
let AllowBoolValue  = new  Map<string, boolean>([
    ["1", true],
    ["0", false],
    ["t", true],
    ["f", false],
    ["T", true],
    ["F", false],
    ["true", true],
    ["false", false],
    ["TRUE", true],
    ["FALSE", false],
    ["True", true],
    ["False", false]
])

function convert(t: KindType, val: string) : any {
    if (t == KindType.String) {
        return val
    } else if (t == KindType.Bool) {
        if (val == null) {
            return true
        }
        let r = AllowBoolValue.has(val)
        if (r === undefined)
            console.log(`wrong bool value: ${val}`)
        else
            return AllowBoolValue.get(val);
    } else if (t == KindType.Number) {
        return Number.parseFloat(val)
    } else {
        console.log("unknow type", t)
    }
}

enum KindType {
    Bool,
    String,
    Number
}

class Entry {
    constructor(public HasDefault: boolean, public Default: any,
               public Usage: string, public Kind: KindType, public Value: any[]) {
    }
}

let Defined = new Map<string, Entry>();
export let Args = new Array<string>();

export function Parse(...args : string[]) {
    let stop = false
    for (let i = 0; i < args.length; i++) {
        let arg = args[i]
        if (stop || arg.slice(0, 1) != "-") {
            Args.push(arg)
        } else if (arg == "--") {
            stop = true
        } else if (arg == "-") {
            stop = true
            Args.push(arg)
        } else {
            let ms = arg.match(/-(.+)=(.+)/)
            let k, v: string = ""
            if (ms) {
                k = ms[1].toString()
                v = ms[2].toString()
            }
            if (k) {
                // -flag=x
                let entry: Entry | undefined = Defined.get(k)
                if (!entry) {
                    console.log("undefined flag", k)
                } else {
                    if (entry.HasDefault) {
                        entry.Value.pop()
                        entry.HasDefault = false
                    }
                    entry.Value.push(convert(entry.Kind, v))
                }
            } else {
                let ms = arg.match(/-(.+)/)
                if (ms)
                    k = ms[1].toString();
                if (k) {
                    let a = args[i + 1]
                    if (a && a.slice(0, 1) != "-") {
                        // -flag x
                        let entry : Entry | undefined = Defined.get(k)
                        if (!entry) {
                            console.log("undefined flag", k)
                        } else {
                            if (entry.HasDefault) {
                                entry.Value.pop()
                                entry.HasDefault = false
                            }
                            entry.Value.push(convert(entry.Kind, a))
                        }
                        i = i + 1
                    } else {
                        // -flag
                        let entry : Entry | undefined = Defined.get(k)
                        if (!entry) {
                            console.log(`undefined flag ${k}`)
                        } else {
                            if (entry.Kind != KindType.Bool) {
                                console.log(`-${k} need one argument`)
                            }
                            if (entry.HasDefault) {
                                entry.Value.pop()
                                entry.HasDefault = false
                            }
                            entry.Value.push(convert(entry.Kind, ""))
                        }
                    }
                }
            }
        }
    } // end of "for (let i = 0; i < args.length; i++) {"
}

export function Bool(name: string, value: boolean, usage: string) {
    let ptr = [value]
    let e = new Entry(true, value, usage, KindType.Bool, ptr);
    Defined.set(name, e)
    return ptr
}

export function String(name: string, value: string, usage: string) {
    let ptr = [value]
    let e = new Entry(true, value, usage, KindType.String, ptr);
    Defined.set(name, e)
    return ptr
}


export function Number2(name: string, value: number, usage: string) {
    let ptr = [value]
    let e = new Entry(true, value, usage, KindType.Number, ptr);
    Defined.set(name, e)
    return ptr
}

export function PrintDefaults() {
    for (let [k, entry] of Defined) {
        console.log(`-${k} ${entry.Kind}`)
        console.log(`  ${entry.Usage} (default ${entry.Default})`)
    }
}
